crates/ra_hir_def/src/nameres/collector.rs

```diff
@@ -438,35 +443,51 @@ where
         } else {
             match import.path.segments.last() {
                 Some(last_segment) => {
-                    let name = match &import.alias {
-                        Some(ImportAlias::Alias(name)) => name.clone(),
-                        Some(ImportAlias::Underscore) => last_segment.clone(), // FIXME rust-analyzer#2736
-                        None => last_segment.clone(),
+                    let (name, is_named) = match &import.alias {
+                        Some(ImportAlias::Alias(name)) => (name.clone(), true),
+                        Some(ImportAlias::Underscore) => (last_segment.clone(), false), // FIXME rust-analyzer#2736
+                        None => (last_segment.clone(), true),
                     };
-                    log::debug!("resolved import {:?} ({:?}) to {:?}", name, import, def);
+                    log::info!(
+                        "resolved import !!name~{:?} ~import~({:?}) to ~def~{:?}",
+                        name,
+                        import,
+                        def
+                    );
 
                     // extern crates in the crate root are special-cased to insert entries into the extern prelude: rust-lang/rust#54658
                     if import.is_extern_crate && module_id == self.def_map.root {
+                        log::info!(
+                            "resolved EXTERN CRATE import ~name~{:?} ~import~({:?}) to ~def~{:?}",
+                            name,
+                            import,
+                            def
+                        );
                         if let Some(def) = def.take_types() {
                             self.def_map.extern_prelude.insert(name.clone(), def);
                         }
                     }
 
-                    self.update(module_id, &[(name, def)], vis);
+                    self.update(module_id, &[(name, def, is_named)], vis);
                 }
                 None => tested_by!(bogus_paths),
             }
         }
     }
 
-    fn update(&mut self, module_id: LocalModuleId, resolutions: &[(Name, PerNs)], vis: Visibility) {
+    fn update(
+        &mut self,
+        module_id: LocalModuleId,
+        resolutions: &[(Name, PerNs, bool)],
```

Comment:
can this not be `[(Option<Name>, PerNs)]`?

Reply:
resolutions is used only in those two places you named an in the `dump` function (here)[https://github.com/rust-analyzer/rust-analyzer/blob/3f70236965d3c6f03c12f890532f0683248fac2f/crates/ra_hir_def/src/nameres.rs#L235] which according to its doc is only for debugging purposes (but appears in a test). Resolutions is also used in two tests. If it is ok for a resolution to be "unnamed" (i'm talking about the concept) then maybe it would be better to change resolutions and actually create the `Resolution` type.
Regarding the option, I think it makes sense